### PR TITLE
Add Variant serdeProxy support

### PR DESCRIPTION
**Pull request by WebFreak001**
support e.g. deserializing int to Variant!(void, ProxiedInt)

Side-effect: this allows us to easily implement serdeProxyCast, which is
done here and supersedes the PR #22

see test case to see what it actually fixes.

Depends on https://github.com/libmir/mir-algorithm/pull/420

### CODE REVIEW

File: source/mir/ion/conv.d [hunk 1]
```diff
@@ -562,3 +562,35 @@ version(mir_ion_test) unittest
...
+    else static if (isImplicitlyConvertible!(From, To))
+    {
+        To ret = v;
+        return ret;
+    }
+    else
+        return to!To(v);
+}
+
+To proxyTo(To, From)(auto ref scope return From v)
```
Comment (9il) [inline source/mir/ion/conv.d:582]: Please do not introduce a separate function for that logic.

Comment (9il) [reply source/mir/ion/conv.d:582]: Instead, place the logic where it is needed.

File: source/mir/ion/conv.d [hunk 2]
```diff
@@ -562,3 +562,35 @@ version(mir_ion_test) unittest
     ])
         text.text2ion.ion2msgpack.msgpack2ion.ion2text.should == text;
 }
+
+To toProxy(To, From)(auto ref scope return From v)
```
Comment (9il) [inline source/mir/ion/conv.d:566]: Ditto, place the logic where it is needed.

File: source/mir/deser/package.d [hunk 3]
```diff
@@ -442,6 +443,7 @@ template deserializeValue(string[] symbolTable)
         if (!isFirstOrderSerdeType!T)
     {with(params){
         import mir.algebraic: isVariant, isNullable;
+        import mir.serde: ContainsProxied;
```
Comment (9il) [inline source/mir/deser/package.d:446]: Why not `Contains!(staticMap!(serdeGetFinalProxy, Types))`?

File: source/mir/deser/package.d [hunk 4]
```diff
@@ -1018,27 +1037,37 @@ template deserializeValue(string[] symbolTable)
...
                     }
                 }
 
-                static if (contains!long)
+                enum containsIntegral = contains!short
+                    || contains!ushort
+                    || contains!int
+                    || contains!uint
+                    || contains!long
+                    || contains!ulong;
```
Comment (9il) [inline source/mir/deser/package.d:1050]: That should be a separate MR with tests.

File: source/mir/deser/package.d [hunk 5]
```diff
@@ -876,7 +876,26 @@ template deserializeValue(string[] symbolTable)
             }
 
             alias Types = T.AllowedTypes;
-            alias contains = Contains!Types;
+            alias contains = ContainsProxied!Types;
+
+            pragma(inline, true) void setValue(T)(T new_value)
```
Comment (9il) [inline source/mir/deser/package.d:881]: That shouldn't be a common logic. What is your use case?

File: source/mir/deser/package.d (hunk 3 above)
Comment (WebFreak001) [reply source/mir/deser/package.d:446]: good idea, didn't think of that, but probably a lot simpler

File: source/mir/deser/package.d (hunk 5 above)
Comment (WebFreak001) [reply source/mir/deser/package.d:881]: this function is only within the isVariant branch, it's the logic to set a variant to a value that might be using a proxy to a proxy-supported value.

This function is the core functionality of this PR and makes the added example pass.

File: source/mir/ion/conv.d (hunk 1 above)
Comment (WebFreak001) [reply source/mir/ion/conv.d:582]: would it be ok to put them in the ser/deser packages? the same code is reused multiple times, that's why I made functions for the a little more complicated/configurable assignment logic.

Having that in a function means there is just a single place to change to control how proxies are assigned.

File: source/mir/deser/package.d (hunk 5 above)
Comment (9il) [reply source/mir/deser/package.d:881]: It is a too advanced feature for the Dlang's small community. It may make other features like better @nogc support via @serdeScoped harder to implement. Please describe what is your use case. I wonder if we can find another way to support it.

Comment (WebFreak001) [reply source/mir/deser/package.d:881]: do you mean the problem is that this is a local function with scope? I can also make it static and use a ref parameter to take in the value to change.

The function body should have no problem with `@nogc` if using `@serdeProxyCast`. The branch that may potentially allocate is also only compiled in if there is a variant that has types which are proxies to primitives / otherwise unsupported value types of the variant.

Comment (9il) [reply source/mir/deser/package.d:881]: No. I mean that there is unlikely should be any function of any kind. The idea itself is good. However, it may block some other code evolution paths. So, could we replace it with minimal hardcoded `if else` conditions for a minimal amount of types? What types do you need to support?

**Comment by WebFreak001** [comment]
closing this for now as this tries to cover too many use-cases. For me right now I only really need #22, which I wanted to generalize here. Originally the idea was using this PR, #22 can be simulated by using a struct with serdeProxy inside a variant and the following support code:

```d
template MyEnumProxyImpl(T)
{
	@serdeProxy!T
	struct MyEnumProxyImpl
	{
		T value;
		alias value this;

		this(T value)
		{
			this.value = value;
		}
	}
}

alias serdeEnumProxy(T) = serdeProxy!(MyEnumProxyImpl!T);
```
